File: mobile_flutter/lib/core/services/remote_logger.dart
// =============================================================================
// remote_logger.dart
// Envía logs / errores al backend (POST /logs/mobile) de forma asíncrona
// y NO bloqueante. Hace fire-and-forget: si falla, no spamea ni rompe UI.
//
// Errors locales (Logger console) + remotos (backend stdout). Doble visibilidad.
//
// Como leer en el VPS:
//   ssh root@62.146.228.180 "docker logs -f pos_ldmapp_backend 2>&1 | grep MOBILE_LOG"
// =============================================================================

import 'dart:async';
import 'package:dio/dio.dart';
import 'package:flutter/foundation.dart';
import 'package:flutter/widgets.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:logger/logger.dart';
import 'package:pos_mobile/app/constants/api_constants.dart';
import 'package:pos_mobile/core/services/device_info_service.dart';

final remoteLoggerProvider = Provider<RemoteLogger>((ref) {
  return RemoteLogger(
    deviceInfo: ref.watch(deviceInfoServiceProvider),
  );
});

enum LogLevel { debug, info, warning, error, fatal }

class RemoteLogger {
  RemoteLogger({required DeviceInfoService deviceInfo})
      : _device = deviceInfo,
        _sessionId = DateTime.now().millisecondsSinceEpoch.toString(),
        _dio = Dio(
          BaseOptions(
            baseUrl: ApiConstants.baseUrl,
            connectTimeout: const Duration(seconds: 6),
            receiveTimeout: const Duration(seconds: 6),
            sendTimeout: const Duration(seconds: 6),
            contentType: 'application/json',
            headers: const {'Accept': 'application/json'},
          ),
        ),
        _local = Logger(
          printer: PrettyPrinter(
            methodCount: 0,
            errorMethodCount: 6,
            lineLength: 100,
            printEmojis: true,
            dateTimeFormat: DateTimeFormat.onlyTimeAndSinceStart,
          ),
        );

  final DeviceInfoService _device;
  final Dio _dio;
  final Logger _local;
  final String _sessionId;

  String? _userId;
  String? _empresaId;
  String? _currentRoute;

  /// Setea el contexto de sesión actual. Llamar después del login y en logout.
  void setSession({String? userId, String? empresaId}) {
    _userId = userId;
    _empresaId = empresaId;
  }

  /// Setea la ruta actual (útil para saber dónde ocurrió el error).
  void setRoute(String? route) {
    _currentRoute = route;
  }

  void debug(String message, {Map<String, Object?>? extra}) =>
      _send(LogLevel.debug, message, extra: extra);

  void info(String message, {Map<String, Object?>? extra}) =>
      _send(LogLevel.info, message, extra: extra);

  void warning(String message, {Map<String, Object?>? extra}) =>
      _send(LogLevel.warning, message, extra: extra);

  void error(
    String message, {
    Object? error,
    StackTrace? stackTrace,
    Map<String, Object?>? extra,
  }) =>
      _send(
        LogLevel.error,
        message,
        error: error,
        stackTrace: stackTrace,
        extra: extra,
      );

  void fatal(
    String message, {
    Object? error,
    StackTrace? stackTrace,
    Map<String, Object?>? extra,
  }) =>
      _send(
        LogLevel.fatal,
        message,
        error: error,
        stackTrace: stackTrace,
        extra: extra,
      );

  void _send(
    LogLevel level,
    String message, {
    Object? error,
    StackTrace? stackTrace,
    Map<String, Object?>? extra,
  }) {
    // 1. Log local (siempre, también en release para que el dev vea).
    switch (level) {
      case LogLevel.fatal:
      case LogLevel.error:
        _local.e(message, error: error, stackTrace: stackTrace);
      case LogLevel.warning:
        _local.w(message);
      case LogLevel.info:
        _local.i(message);
      case LogLevel.debug:
        _local.d(message);
    }

    // 2. Envio remoto fire-and-forget.
    unawaited(_postRemote(level, message, error, stackTrace, extra));
  }

  Future<void> _postRemote(
    LogLevel level,
    String message,
    Object? error,
    StackTrace? stackTrace,
    Map<String, Object?>? extra,
  ) async {
    try {
      final body = <String, dynamic>{
        'level': level.name,
        'message': message.length > 500 ? message.substring(0, 500) : message,
        if (error != null)
          'error': error.toString().length > 2000
              ? error.toString().substring(0, 2000)
              : error.toString(),
        if (stackTrace != null)
          'stackTrace': stackTrace.toString().length > 8000
              ? stackTrace.toString().substring(0, 8000)
              : stackTrace.toString(),
        if (_currentRoute != null) 'route': _currentRoute,
        if (_userId != null) 'userId': _userId,
        if (_empresaId != null) 'empresaId': _empresaId,
        'sessionId': _sessionId,
        'timestamp': DateTime.now().toIso8601String(),
        'device': _device.snapshot.toJson(),
        if (extra != null && extra.isNotEmpty) 'extra': extra.toString(),
      };

      await _dio.post<dynamic>('/logs/mobile', data: body);
    } catch (e) {
      // Falló el envío remoto: NO bucle infinito ni toast. Solo log local.
      if (kDebugMode) {
        // ignore: avoid_print
        print('[RemoteLogger] no se pudo enviar: $e');
      }
    }
  }
}

/// NavigatorObserver que actualiza la ruta actual en el RemoteLogger.
/// Así cualquier error tiene el contexto de dónde ocurrió.
class RemoteLoggerRouteObserver extends NavigatorObserver {
  RemoteLoggerRouteObserver(this._logger);
  final RemoteLogger _logger;

  void _update(Route<dynamic>? route) {
    final name = route?.settings.name;
    if (name != null) _logger.setRoute(name);
  }

  @override
  void didPush(Route<dynamic> route, Route<dynamic>? previousRoute) =>
      _update(route);

  @override
  void didReplace({Route<dynamic>? newRoute, Route<dynamic>? oldRoute}) =>
      _update(newRoute);

  @override
  void didPop(Route<dynamic> route, Route<dynamic>? previousRoute) =>
      _update(previousRoute);
}

File: mobile-supabase/src/services/remote-logger.ts
// =============================================================================
// remote-logger.ts — Envía logs/errores al backend para diagnóstico remoto.
// Ver desde el VPS: ssh root@62.146.228.180 "docker logs -f pos_ldmapp_backend | grep MOBILE_LOG"
// =============================================================================

import { Platform } from 'react-native';
import * as Application from 'expo-application';
import * as Device from 'expo-device';
import * as Localization from 'expo-localization';

const ENDPOINT = 'https://api-pos.ldmapp.com/api/v1/logs/mobile';
const SESSION_ID = `${Date.now()}-${Math.random().toString(36).slice(2, 8)}`;

type Level = 'debug' | 'info' | 'warning' | 'error' | 'fatal';

const ctx: { userId?: string; empresaId?: string } = {};
let currentRoute: string | undefined;

const device = {
  platform: Platform.OS,
  osVersion: Platform.Version?.toString() ?? '',
  model: Device.modelName ?? '',
  appVersion: Application.nativeApplicationVersion ?? '0.1.0',
  buildNumber: Application.nativeBuildVersion ?? '1',
  locale: Localization.getLocales?.()[0]?.languageTag ?? 'es-PE',
};

export function setLoggerSession(opts: { userId?: string; empresaId?: string }) {
  ctx.userId = opts.userId;
  ctx.empresaId = opts.empresaId;
}

export function clearLoggerSession() {
  ctx.userId = undefined;
  ctx.empresaId = undefined;
}

export function setLoggerRoute(route: string) {
  currentRoute = route;
}

function truncate(s: unknown, max: number) {
  const str = typeof s === 'string' ? s : String(s);
  return str.length > max ? str.slice(0, max) : str;
}

async function send(level: Level, msg: string, err?: unknown, stack?: string, extra?: Record<string, unknown>) {
  if (__DEV__) {
    const fn = level === 'fatal' || level === 'error' ? 'error' : level === 'warning' ? 'warn' : 'log';
    // eslint-disable-next-line no-console
    console[fn]('[REMOTE_LOG]', msg, err ?? '');
  }
  try {
    const body: Record<string, unknown> = {
      level,
      message: truncate(msg, 500),
      sessionId: SESSION_ID,
      timestamp: new Date().toISOString(),
      device,
    };
    if (err) body.error = truncate(err instanceof Error ? err.message : String(err), 2000);
    if (stack) body.stackTrace = truncate(stack, 8000);
    if (currentRoute) body.route = currentRoute;
    if (ctx.userId) body.userId = ctx.userId;
    if (ctx.empresaId) body.empresaId = ctx.empresaId;
    if (extra) body.extra = truncate(JSON.stringify(extra), 2000);

    await fetch(ENDPOINT, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(body),
    });
  } catch {
    // Fire and forget — no debe romper la app
  }
}

export const remoteLogger = {
  debug: (m: string, e?: Record<string, unknown>) => send('debug', m, undefined, undefined, e),
  info: (m: string, e?: Record<string, unknown>) => send('info', m, undefined, undefined, e),
  warning: (m: string, e?: Record<string, unknown>) => send('warning', m, undefined, undefined, e),
  error: (m: string, err?: unknown, extra?: Record<string, unknown>) => {
    const s = err instanceof Error ? err.stack : undefined;
    return send('error', m, err, s, extra);
  },
  fatal: (m: string, err?: unknown, extra?: Record<string, unknown>) => {
    const s = err instanceof Error ? err.stack : undefined;
    return send('fatal', m, err, s, extra);
  },
};

export function installGlobalErrorHandlers() {
  const g = globalThis as unknown as {
    ErrorUtils?: {
      getGlobalHandler?: () => (e: Error, fatal?: boolean) => void;
      setGlobalHandler?: (h: (e: Error, fatal?: boolean) => void) => void;
    };
  };
  const prev = g.ErrorUtils?.getGlobalHandler?.();
  g.ErrorUtils?.setGlobalHandler?.((err, isFatal) => {
    remoteLogger[isFatal ? 'fatal' : 'error']('GlobalJSError', err);
    prev?.(err, isFatal);
  });
}
